rpc_state_reader_sn_api/src/lib.rs

```diff
@@ -0,0 +1,994 @@
+///
+/// This implementation is uses HTTP requests to call the RPC endpoint,
+/// using Infura.
+/// In order to use it an Infura API key is necessary.
+pub struct RpcState {
+    /// Enum with one of the supported Infura chains/
+    chain: RpcChain,
+    /// Infura API key.
+    api_key: String,
+    /// Struct that holds information on the block where we are going to use to read the state.
+    block: BlockValue,
+}
+
+#[derive(Debug, Error)]
+enum RpcError {
+    #[error("RPC call failed with error: {0}")]
+    RpcCall(String),
+    #[error("Request failed with error: {0}")]
+    Request(String),
+    #[error("Failed to cast from: {0} to: {1} with error: {2}")]
+    Cast(String, String, String),
+}
+
+/// [`BlockValue`] is an Enum that represent which block we are going to use to retrieve information.
+#[allow(dead_code)]
+pub enum BlockValue {
+    /// String one of: ["latest", "pending"]
+    Tag(serde_json::Value),
+    /// Integer
+    Number(serde_json::Value),
+    /// String with format: 0x{felt252}
+    Hash(serde_json::Value),
+}
+
+impl BlockValue {
+    fn to_value(&self) -> serde_json::Value {
+        match self {
+            BlockValue::Tag(block_tag) => block_tag.clone(),
+            BlockValue::Number(block_number) => json!({ "block_number": block_number }),
+            BlockValue::Hash(block_hash) => json!({ "block_hash": block_hash }),
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct RpcResponseContractClass {
+    result: SNContractClass,
+}
+
+// We use this new struct to cast the string that contains a [`StarkFelt`] in hex to a [`StarkFelt`]
+struct StarkFeltHex;
+
+impl<'de> DeserializeAs<'de, StarkFelt> for StarkFeltHex {
+    fn deserialize_as<D>(deserializer: D) -> Result<StarkFelt, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let bytes: PrefixedBytesAsHex<32> = PrefixedBytesAsHex::deserialize(deserializer)?;
+        Ok(bytes.try_into().unwrap())
+    }
+}
+
+#[serde_as]
+#[derive(Debug, Deserialize)]
+struct RpcResponseFelt {
+    #[serde_as(as = "StarkFeltHex")]
+    result: StarkFelt,
+}
+
+pub struct RpcBlockInfo {
+    /// The sequence number of the last block created.
+    pub block_number: BlockNumber,
+    /// Timestamp of the beginning of the last block creation attempt.
+    pub block_timestamp: BlockTimestamp,
+    /// The sequencer address of this block.
+    pub sequencer_address: ContractAddress,
+}
+
+type RpcResponse = ureq::Response;
+
+impl RpcState {
+    pub fn new(chain: RpcChain, block: BlockValue) -> Self {
+        if env::var("INFURA_API_KEY").is_err() {
+            dotenv().expect("Missing .env file");
+        }
+        Self {
+            chain,
+            api_key: env::var("INFURA_API_KEY")
+                .expect("Missing API Key in environment: INFURA_API_KEY"),
+            block,
+        }
+    }
+
+    fn rpc_call<T: for<'a> Deserialize<'a>>(
+        &self,
+        params: &serde_json::Value,
+    ) -> Result<T, RpcError> {
```

Comment:
Just to know,
won't be better to return a `Result<RpcResponse,..>`?

Reply:
There are cases in which it's useful to deserialize into a `serde_json::Value`, e.g. `get_transaction()`.